moveit_ros/moveit_servo/src/pose_tracking.cpp

```diff
@@ -0,0 +1,304 @@
+ *     with the distribution.
+ *   * Neither the name of PickNik LLC nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#include "moveit_servo/pose_tracking.h"
+
+namespace
+{
+constexpr char LOGNAME[] = "pose_tracking";
+constexpr double DEFAULT_LOOP_RATE = 100;     // Hz
+constexpr double ROS_STARTUP_WAIT = 10;       // sec
+constexpr double DEFAULT_POSE_TIMEOUT = 0.1;  // sec
+}
+
+namespace moveit_servo
+{
+PoseTracking::PoseTracking(const planning_scene_monitor::PlanningSceneMonitorPtr& planning_scene_monitor,
+                           const std::string& parameter_ns)
+  : planning_scene_monitor_(planning_scene_monitor)
+  , loop_rate_(DEFAULT_LOOP_RATE)
+  , transform_listener_(transform_buffer_)
+  , stop_requested_(false)
+  , parameter_ns_(parameter_ns)
+  , angular_error_(0)
+{
+  readROSParams();
+
+  robot_model_ = planning_scene_monitor_->getRobotModel();
+  joint_model_group_ = robot_model_->getJointModelGroup(move_group_name_);
+
+  // Initialize PID controllers
+  initializePID(x_pid_config_, cartesian_position_pids_);
+  initializePID(y_pid_config_, cartesian_position_pids_);
+  initializePID(z_pid_config_, cartesian_position_pids_);
+  initializePID(angular_pid_config_, cartesian_orientation_pids_);
+
+  // Use the C++ interface that Servo provides
+  servo_ = std::make_unique<moveit_servo::Servo>(nh_, planning_scene_monitor_, parameter_ns_);
+  servo_->start();
+
+  // Connect to Servo ROS interfaces
+  std::string target_pose_topic = "/" + parameter_ns_ + "/target_pose";
+  target_pose_sub_ =
+      nh_.subscribe<geometry_msgs::PoseStamped>(target_pose_topic, 1, &PoseTracking::targetPoseCallback, this);
+
+  // Publish outgoing twist commands to the Servo object
+  twist_stamped_pub_ =
+      nh_.advertise<geometry_msgs::TwistStamped>(servo_->getParameters().cartesian_command_in_topic, 1);
+}
+
+int8_t PoseTracking::moveToPose(const Eigen::Vector3d& positional_tolerance, const double angular_tolerance)
```

Comment:
I am a bit confused by the `PoseTracking` interface. It seems the only way to tell `PoseTracking` to start going to a pose is by directly calling this function, and the only way to set the pose is by publishing it via ROS. I also think the first bit of this function means you have to publish the pose within a specific time of calling the function.

Would it make sense to allow the C++ call here to include a pose to move to (so the call is `tracker_->moveToPose(pos_tol, ang_tol, pose)`)?

Other thoughts: the `moveToPose` could be made into a timer callback that ensure the tracker is always moving to the current pose target (set via publishing). Then there are `start()` and `stop()` interfaces for controlling that timer, as in ServoCalcs/CollisionCheck

Reply:
The reason I want to publish the target pose on a ROS topic is to update it in real time. You can't do that with a call like this:

`tracker_->moveToPose(pos_tol, ang_tol, pose)`

I'll think about stuffing this in a timer callback

Reply:
@AdamPettinger (and @tylerjw) can you be more specific about the benefits of moving this into a Timer callback?